app/models/client_opportunity_match_contact.rb

```diff
@@ -16,14 +16,46 @@ class ClientOpportunityMatchContact < ApplicationRecord
 
   acts_as_paranoid
 
+  scope :for_contact_type, ->(contact_type) do
+    contact_type_column = ClientOpportunityMatchContact.column_for(contact_type)
+    where(ClientOpportunityMatchContact.arel_table[contact_type_column.to_sym].eq(true))
+  end
+
   def self.text_search(text)
     return none unless text.present?
 
     contact_matches = Contact.where(
-      Contact.arel_table[:id].eq(arel_table[:contact_id])
+      Contact.arel_table[:id].eq(arel_table[:contact_id]),
     ).text_search(text).arel.exists
 
     where(contact_matches)
   end
 
+  def self.contact_types
+    [
+      :shelter_agency,
+      :client,
+      :dnd_staff,
+      :housing_subsidy_admin,
+      :ssp,
+      :hsp,
+      :do,
+    ].freeze
+  end
+
+  def self.contact_type_columns
+    contact_types.map { |t| ["#{t}_contacts".to_sym, t] }.to_h
+  end
+
+  def self.column_for(contact_type)
+    contact_type_columns[contact_type] || contact_type
+  end
+
+  def self.join_contact_methods
+    contact_types.map { |t| ["#{t}_contacts".to_sym, "#{t}_join_contacts".to_sym] }.to_h
+  end
+
+  def self.join_method_for(contact_type)
+    join_contact_methods[contact_type] || contact_type
+  end
```

Comment:
This feels good.  Thanks.

Reply:
Do we want to clean up the client, program, and subprogram lists too, or do you think its better to leave them as is for now?

Reply:
I just pushed a new commit with the list updates for the additional contact files